doc/designs/application-specific-passwords.md

```diff
@@ -0,0 +1,243 @@
+convenience in the case that the user can no longer trust a password used on
+one of his devices. Namely, he can revoke or change only that password, without
+needing to revoke or change his other passwords, including the primary password.
+Access control extension would add significant security benefit to this - the
+intruder's access to the user's account might be more or less limited
+according to what permissions is the compromised password assigned.
+
+Resolves: https://pagure.io/freeipa/issue/4510
+How Google does it: https://support.google.com/accounts/answer/185833?hl=en
+
+## Use cases
+
+### Use with password-only authentication
+
+The user can set multiple passwords per account. Each password can be then used
+for authenticating on only one device. Various passwords can as well be used to
+authenticate to various services or their sets. These divisions can even be
+combined, thus - in the most extreme case - having one password for
+authenticating to one specific service on one particular device. Granularity of
+such divisions will be fully up to user's discretion.
+
+### Use as a part of 2FA (OTP)
+
+The user could enter an application-specific password instead of the primary
+one, and then an OTP, allowing application-specific passwords to be used as
+a part of 2FA. Although 2FA reduces the risk of account hijacking, the benefits
+of application-specific passwords still improve security in this case as well.
+
+### Use as a fallback mechanism when 2FA is activated, but the client
+    application does not support it
+
+The user would enter only an application-specific password, enabling to use 2FA
+for account management as well as services that support it, while allowing
+applications that do not support it to work.
+
+### Use as part of RADIUS and Hardened Password authentication mechanisms
+
+As both of these authentication mechanisms include sending a password for
+authentication, they could also be used with application-specific passwords.
+As pure RADIUS provides only weak protection of the sent password, it might
+be even more desirable than in the case of traditional password authentication.
+
+## Design
+
+### Multiple passwords for a single user and related changes
+
+If the attribute for password in the user schema is not multi-valued, it must
+be modified to be. The user management must be extended to facilitate multiple
+passwords for a single user. This also includes password reset functionality -
+the user will be able to specify what password to reset. The user account
+status will display information for all its passwords.
+
+It is worth considering adding another attribute to the user schema, which will
+hold labels for all application-specific passwords, so that these labels can be
+displayed in the WebUI and used in the relevant commands in CLI, increasing
+the usability of these new functionality. The user will not need to remember
+that e.g. password no. 4 is for access to email on a smartphone.
```

Comment:
there is no ordering in multivalued attribute, and I am generally against changing the nature of the userPassword attribute. It will cause trouble.
You also need to consider if you want to allow application specific passwors via kerberos or limit them to LDAP basic authentication. I can see value both ways so it needs to be analyzed. Also need to analyze whether sub accounts would be created for krb5 (user/appname@REALM) for example, or other schemes need to be empoyed, ahving multiple valid krb5 keys for the same principal is not trivial.

Reply:
Then I think It would be best to store app-specific passwords in a new standalone attribute `userappspecificpassword` with - for each pair - the key being the label and the value being the password itself. This would also logically separate them from the primary password.

Reply:
As for app-specific passwords for Kerberos, it also depends if we want to be able to give permissions to them on par with the primary password, that means including user management, as it requires a Kerberos ticket. Even if not, app-specific passwords for Kerberos may still be useful. But of course, I will analyze how it could be done and whether it is worth it.

Reply:
FYI, ``userPassword`` is a multi-valued octet-string, https://tools.ietf.org/html/rfc4519#section-2.41

Reply:
But this is for LDAP in general, in `ipaserver/plugins/baseuser.py` it is
```
Password('userpassword?',
```
This is a specification of a single-valued attribute that can also be empty. In `ipalib/parameters`.py I see that Password type inherits from Str, which in turn inherits from Data. I can't see how the `userpassword` attribute in FreeIPA can hold multiple values now (I am still learning and may not know about or understand something).

Reply:
Or is it always multi-valued on the LDAP level, even if FreeIPA does not specify it as multi-valued?

Reply:
The LDAP standard for the userPassword attribute allows for multiple entries. IPA limits the userPassword to one entry only. The IPA restriction could be lifted. But that wouldn't be beneficial for your proposal. Attributes are unordered collections. There would be no easy way to distinguish between the main password and app passwords. It might be possible to use tagged attributes, but that might break applications.

Reply:
It doesn't really matter that the attribute is multivalued, most applications dealing with that attribute expect a single entry and would be confused and fail in non-obvious ways if multiple values would be provided. So we should definitely not extend the use of this standard attribute.